### ISSUE TITLE
ouroboros exception when running 

### ISSUE THREAD

**Issue by bilago** [pyouroboros/ouroboros#226]
I've had ouroboros installed for about a week now on Ubuntu 18.10 and I've had to remove the ouroboros image and reinstall it twice now for this error. Once set up again it runs fine the first time, then after that I get this exception. Please advise:

`2019-02-17 16:40:25 : ERROR : base : Job "Interval container update for unix://var/run/docker.sock (trigger: interval[0:05:00], next run at: 2019-02-17 16:45:25 EST)" raised an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
    updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
    if current_image.id != latest_image.id:
AttributeError: 'NoneType' object has no attribute 'id'
2019-02-17 16:45:25 : ERROR : base : Job "Interval container update for unix://var/run/docker.sock (trigger: interval[0:05:00], next run at: 2019-02-17 16:50:25 EST)" raised an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
    updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
    if current_image.id != latest_image.id:
AttributeError: 'NoneType' object has no attribute 'id'
2019-02-17 16:50:25 : ERROR : base : Job "Interval container update for unix://var/run/docker.sock (trigger: interval[0:05:00], next run at: 2019-02-17 16:55:25 EST)" raised an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
    updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
    if current_image.id != latest_image.id:
AttributeError: 'NoneType' object has no attribute 'id'
2019-02-17 16:55:25 : ERROR : base : Job "Interval container update for unix://var/run/docker.sock (trigger: interval[0:05:00], next run at: 2019-02-17 17:00:25 EST)" raised an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
    updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
    if current_image.id != latest_image.id:
AttributeError: 'NoneType' object has no attribute 'id'
2019-02-17 18:02:58 : INFO : ouroboros : Version: 1.2.1-master`

**Comment by DirtyCajunRice** [comment]
please post your config settings (first line in debug output on start)

**Comment by bilago** [comment]
Log level is currently at info, i'll change it to debug, however reinstalling the image will fix the problem for now so it might be a bit before I can replicate the error message above.

How i installed the image:


```
docker create \
--name=ouroboros \
-e CLEANUP=true \
-e INTERVAL=300 \
-e LOG_LEVEL=info \
-e SELF_UPDATE=true \
-e IGNORE=mongo \
-e IGNORE=influxdb \
-e IGNORE=postgres \
-e IGNORE=mariadb \
-e TZ=America/New_York \
-v /var/run/docker.sock:/var/run/docker.sock \
-v /home/bilago/.docker:/root/.docker \
pyouroboros/ouroboros
```

**Comment by bilago** [comment]
config settings with debug enabled:


`2019-02-17 18:15:48 : DEBUG : ouroboros : Ouroboros configuration:
 {
'interval': 300, 
'docker_sockets': ['unix://var/run/docker.sock'], 
'ignore': ['mariadb'], 
'log_level': 'debug', 
'prometheus_addr': None, 
'prometheus_port': None, 
'cleanup': True, 
'influx_url': None,
 'influx_port': None, 
'influx_username': None, 
'influx_password': None, 
'self_update': True, 
'hostname': '5b7980c51586', 
'docker_tls_verify': True
}`

**Comment by DirtyCajunRice** [comment]
im almost certain you are on 1.2.0. this was fixed in 1.2.1

**Comment by bilago** [comment]
According to the logs:

`INFO : ouroboros : Version: 1.2.1-master`

**Comment by DirtyCajunRice** [comment]
can you completely remove the container and image(s) for ouroboros and pull a clean fresh image? the error you are getting shouldnt be possible with the current code

**Comment by bilago** [comment]
did the following

- docker stop ouroboros
- docker remove ouroboros
- docker rmi pyouroboros/ouroboros
- docker pull pyouroboros/ouroboros
- installed using my docker create command

I'll update this bug report with my results as it might take some time before the error occurs again

**Comment by DirtyCajunRice** [comment]
As a side note, multiple ignores dont work like that. An environment variable cannot be set more than once, so the last ignore is the only ignore that will work. Instead do `-e IGNORE=mongo influxdb postgres mariadb`

**Comment by tonobo** [comment]
I've experienced the same issue by accessing our private registry. 

```
2019-02-19 10:09:53 : DEBUG : dockerclient : Checking tag: registry.example.com/path/to/docker:experimental
2019-02-19 10:09:53 : DEBUG : connectionpool : http://localhost:None "GET /v1.35/distribution/registry.example.com/path/to/docker:experimental/json HTTP/1.1" 500 135
```

docker-compose.yml

```yaml
  ouroboros:
    container_name: ouroboros
    hostname: ouroboros
    image: pyouroboros/ouroboros
    environment:
      CLEANUP: "true"
      INTERVAL: 30
      LOG_LEVEL: debug
      SELF_UPDATE: "true"
      LABEL_ENABLE: "true"
      LABELS_ONLY: "true"
      DRY_RUN: "true"
      REPO_USER: gitlab-user
      REPO_PASS: secret
      TZ: Europe/Berlin
    restart: unless-stopped
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /root/.docker/config.json:/root/.dockercfg
      - /root/.docker/config.json:/root/.docker/config.json
```

/root/.docker/config.json:
```
{
	"auths": {
		"registry.example.com": {
			"auth": "secret"
		}
	},
	"HttpHeaders": {
		"User-Agent": "Docker-Client/18.09.2 (linux)"
	}
}
```

**Comment by zegervdv** [comment]
I've got the same issue too:
```
ouroboros | 2019-02-20 10:07:44 : INFO : ouroboros : Version: 1.2.1-master
ouroboros | 2019-02-20 10:07:46 : ERROR : base : Job "Initial run interval container update for unix://var/run/docker.sock (trigger: date[2019-02-20 10:07:44 UTC], next run at: 2019-02-20 10:07:44 UTC)" raised an exception
ouroboros | Traceback (most recent call last): 
ouroboros |   File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
ouroboros |     retval = job.func(*job.args, **job.kwargs)
ouroboros |   File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
ouroboros |     updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
ouroboros |   File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
ouroboros |     if current_image.id != latest_image.id:
ouroboros | AttributeError: 'NoneType' object has no attribute 'id' 
```

I am also using a private registry.

Configuration:
```
ouroboros:
  image: pyouroboros/ouroboros:latest
  container_name: ouroboros
  restart: unless-stopped
  environment: 
    - DATA_EXPORT=prometheus
    - PROMETHEUS_ADDR=0.0.0.0
    - PROMETHEUS_PORT=8000
  volumes:
    - /var/run/docker.sock:/var/run/docker.sock 
    - ./config.json:/root/.docker/config.json:ro
  command: --label-enable --labels-only --interval 30
```

**Comment by zegervdv** [comment]
I've set a breakpoint and found the `latest_image` is `None`.

It seems the pulling of the current image fails, which is swallowed by the `try` block.

**Comment by bilago** [comment]
I left the container running since we last chatted, and I'm now getting the error again:

`2019-02-20 17:56:22 : DEBUG : connectionpool : http://localhost:None "POST /v1.35/images/create?tag=latest&fromImage=pyouroboros%2Fouroboros HTTP/1.1" 500 160
2019-02-20 17:56:22 : ERROR : base : Job "Interval container update for unix://var/run/docker.sock (trigger: interval[0:05:00], next run at: 2019-02-20 18:01:22 EST)" raised an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
    retval = job.func(*job.args, **job.kwargs)
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 276, in update
    updateable, depends_on_containers, hard_depends_on_containers = self.socket_check()
  File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 241, in socket_check
    if current_image.id != latest_image.id:
AttributeError: 'NoneType' object has no attribute 'id'`

**Comment by circa10a** [comment]
@bilago what version are you running?

**Comment by bilago** [comment]
'INFO : ouroboros : Version: 1.2.1-master'

**Comment by cbirkenbeul** [comment]
I've got the same error. Also did a clean pull from the new 1.2.1-master. Starting it via compose

`  ouroboros:
    container_name: ouroboros
    hostname: ouroboros
    image: pyouroboros/ouroboros:latest
    environment:
      - CLEANUP=true
      - INTERVAL=300
      - LOG_LEVEL=info
      - SELF_UPDATE=true
      - TZ=Europe/Berlin
      - NOTIFIERS=slack://token1/token2/token3/#monitoring
    restart: unless-stopped
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock`

Error output:
`ouroboros    | 2019-02-22 12:45:36 : INFO : ouroboros : Version: 1.2.1-master
ouroboros    | 2019-02-22 12:45:36 : WARNING : NotifySlack : No user was specified; using apprise.
ouroboros    | 2019-02-22 12:45:37 : INFO : NotifySlack : Sent Slack notification.
ouroboros    | 2019-02-22 12:46:59 : INFO : dockerclient : pat-db will be updated
ouroboros    | 2019-02-22 12:47:02 : ERROR : dockerclient : Could not delete old image for pat-db, Error: 409 Client Error: Conflict ("conflict: unable to delete 1571059e8458 (cannot be forced) - image is being used by running container 9f7e3e1673e6")
ouroboros    | 2019-02-22 12:47:02 : INFO : dockerclient : teqqy-de-ghost will be updated
ouroboros    | 2019-02-22 12:47:09 : INFO : dockerclient : unifiguide-app will be updated
ouroboros    | 2019-02-22 12:47:10 : ERROR : base : Job "Initial run interval container update for unix://var/run/docker.sock (trigger: date[2019-02-22 12:45:36
CET], next run at: 2019-02-22 12:45:36 CET)" raised an exception
ouroboros    | Traceback (most recent call last):
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/apscheduler/executors/base.py", line 125, in run_job
ouroboros    |     retval = job.func(*job.args, **job.kwargs)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 301, in update
ouroboros    |     self.recreate(container, latest_image)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/pyouroboros/dockerclient.py", line 117, in recreate
ouroboros    |     network.connect(**new_network_config)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/models/networks.py", line 57, in connect
ouroboros    |     container, self.id, *args, **kwargs
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/utils/decorators.py", line 19, in wrapped
ouroboros    |     return f(self, resource_id, *args, **kwargs)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/api/network.py", line 242, in connect_container_to_network
ouroboros    |     ipv6_address=ipv6_address, link_local_ips=link_local_ips
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/api/container.py", line 649, in create_endpoint_config
ouroboros    |     return EndpointConfig(self._version, *args, **kwargs)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/types/networks.py", line 17, in __init__
ouroboros    |     self["Links"] = normalize_links(links)
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/utils/utils.py", line 447, in normalize_links
ouroboros    |     return ['{0}:{1}'.format(k, v) if v else k for k, v in sorted(links)]
ouroboros    |   File "/usr/local/lib/python3.7/site-packages/docker/utils/utils.py", line 447, in <listcomp>
ouroboros    |     return ['{0}:{1}'.format(k, v) if v else k for k, v in sorted(links)]
ouroboros    | ValueError: too many values to unpack (expected 2)`

**Comment by circa10a** [comment]
@cbirkenbeul  your error is due to using docker links instead of using docker networks. Docker-py doesn't handle links very well. Could you try using a network instead?

**Comment by circa10a** [comment]
The issue here isn't the private registry, i'm unable to reproduce from my end.

**Comment by circa10a** [comment]
@DirtyCajunRice thoughts on closing this due to recent merge?

**Comment by DirtyCajunRice** [comment]
sounds good. on phone so you gotta close it

**Comment by circa10a** [comment]
Fixed in #237
